File: ResourceManager/Models/Activity.cs
using ResourceManager.Models.Abstractions;
using System;
using System.Collections.Generic;
using System.Linq;

namespace ResourceManager.Models
{
    public class Activity : BaseEntity, ITags, IEquatable<Activity>, IComparable<Activity>
    {
        public string Title { get; set; } = string.Empty;
        public DateTime Date { get; set; }
        public DateTime StartTime { get; set; }
        public DateTime EndTime { get; set; }
        public int Duration { get; set; }
        public string Metric { get; set; } = string.Empty;
        public int Quantity { get; set; }
        public string Output { get; set; } = string.Empty;
        public string Note { get; set; } = string.Empty;
        public int UserID { get; set; }
        public User User { get; set; }
        public IList<ActivityTag> ActivityTags { get; set; } = new List<ActivityTag>();

        public string TagString => string.Join(", ", ActivityTags.Select(x => x.Tag.Name));

        public override string GetCaption()
        {
            return string.Format("{0} with Title {1}", nameof(Activity), Title);
        }

        #region IEquatable<Activity>
        public bool Equals(Activity? other)
        {
            if (ReferenceEquals(null, other)) return false;

            return Title.Equals(other.Title) && Date.Equals(other.Date);
        }
        #endregion

        #region IComparable<Activity>
        public int CompareTo(Activity? other)
        {
            if (ReferenceEquals(null, other)) return -1;
            int result = Date.CompareTo(other.Date);
            if (result == 0)
            {
                return Title.CompareTo(other.Title);
            }
            return result;
        }
        #endregion
    }
}

File: ResourceManager/Models/Activity.vb
Imports ResourceManager.Models.Abstractions
Imports System
Imports System.Collections.Generic
Imports System.Linq

Namespace ResourceManager.Models
    Public Class Activity
        Inherits BaseEntity
        Implements ITags, IEquatable(Of Activity), IComparable(Of Activity)
        Public Property Title As String = String.Empty
        Public Property [Date] As Date
        Public Property StartTime As Date
        Public Property EndTime As Date
        Public Property Duration As Integer
        Public Property Metric As String = String.Empty
        Public Property Quantity As Integer
        Public Property Output As String = String.Empty
        Public Property Note As String = String.Empty
        Public Property UserID As Integer
        Public Property User As User
        Public Property ActivityTags As ICollection(Of ActivityTag)

        Public ReadOnly Property TagString As String Implements ITags.TagString
            Get
                Return String.Join(", ", ActivityTags.Select(Function(x) x.Tag.Name))
            End Get
        End Property

        Public Overrides Function GetCaption() As String
            Return String.Format("{0} with Title {1}", NameOf(Activity), Title)
        End Function

#Region "IEquatable<Activity>"
        Public Shadows Function Equals(ByVal other As Activity) As Boolean Implements IEquatable(Of Activity).Equals
            If ReferenceEquals(Nothing, other) Then Return False

            Return Title.Equals(other.Title) AndAlso [Date].Equals(other.Date)
        End Function
#End Region

#Region "IComparable<Activity>"
        Public Function CompareTo(ByVal other As Activity) As Integer Implements IComparable(Of Activity).CompareTo
            If ReferenceEquals(Nothing, other) Then Return -1
            Dim result = [Date].CompareTo(other.Date)
            If result = 0 Then
                Return Title.CompareTo(other.Title)
            End If
            Return result
        End Function
#End Region
    End Class
End Namespace
